### PR TITLE
Remote Signing

### LINKED ISSUE #37: Merge Remote Signing Branch into Main
To prepare for the Alpha.1 release, we'll

- [x] Complete the coding in the remote signing branch
- [x] Complete the test coverage, with passing tests
- [x] Create a tag [`v0.0.0-sign1-alpha.0`](https://github.com/veraison/go-cose/releases/tag/v0.0.0-sign1-alpha.0) and [`v0.0.0-sign1-alpha.0`](https://github.com/veraison/go-cose/tree/v0.0.0-sign1-alpha.0) branch from main, preserving the content prior to the remote signing updates
- [x] Identify PR #51 to merge into main
- [ ] Merge into main

### PR DESCRIPTION
**Pull request by shizhMSFT**
This PR fills the gap to a mature `go-cose` implementation with security improvements and remote-signing extensibility.
Details are available at [DESIGN.md](https://github.com/veraison/go-cose/blob/5bbeb1d739bd2c484ad930d6e0c1b79064c946c4/DESIGN.md).

It resolves #37 and the following highlighted issues:

- #7
- #8
- #9 
- #10 
- #22 
- #23
- #32 
- #36 
- #41 

This PR is co-authored by @qmuntal.

### CODE REVIEW

**Comment by shizhMSFT** [comment]
@thomas-fossati The DCO check fails on 5 commits with error message:
```
Author: Quim Muntal, Committer: GitHub; Expected "Quim Muntal [quimmuntal@gmail.com](mailto:quimmuntal@gmail.com)", but got "qmuntal [qmuntaldiaz@microsoft.com](mailto:qmuntaldiaz@microsoft.com)".
```

How can we resolve it? or should we just set DCO to pass?

/cc @qmuntal

**Comment by SteveLasker** [comment]
unfortunately, bypassing dco bypasses the licensing and introduces the potential for ownership of the code. 
Can you resolve in the branch and resubmit?

**Comment by shizhMSFT** [comment]
Squashed commits with DCO fixed.

**Comment by thomas-fossati** [comment]
LGTM, thanks all for the incredible work.

Should we explicitly tag main before merging?  (I think Steve suggested this some time ago.)

**Comment by SteveLasker** [comment]
awesome work!
Yeah, let’s tag main into a branch. 
Naming is hard. @thomas-fossati, what tag/branch name would be meaningful for you?  If we look 6 months from now, what name would make sense?

File: README.md [hunk 1]
```diff
@@ -1,67 +1,105 @@
-# Why forking Mozilla go-cose?
-
-This fork of [Mozilla go-cose](https://github.com/mozilla-services/go-cose) adds [Sign1](https://tools.ietf.org/html/rfc8152#section-4.2) capabilities to the parent package.  We hope we can reconcile our fork quite quickly.  In the meantime, you can inspect the delta [here](https://github.com/mozilla-services/go-cose/compare/master...veraison:master).
-
 # go-cose
 
-[![GitHub CI](https://github.com/veraison/go-cose/workflows/ci/badge.svg)](https://github.com/veraison/go-cose/actions?query=workflow%3Aci)
-
-[![Coverage Status](https://github.com/veraison/go-cose/workflows/cover%20%E2%89%A589%25/badge.svg)](https://github.com/veraison/go-cose/actions?query=workflow%3A%22cover%20%E2%89%A589%25%22)
+[![go.dev](https://pkg.go.dev/badge/github.com/veraison/go-cose.svg)](https://pkg.go.dev/github.com/veraison/go-cose)
```
Comment (yogeshbdeshpande) [inline README.md:3]: The README.md has been impacted and we do not refer to Mozilla in any way. Should we not mention them at the bottom of the README.md for now, till we have the pending items with them resolved ?

**Comment by thomas-fossati** [comment]
> awesome work!
> Yeah, let’s tag main into a branch.
> Naming is hard. @thomas-fossati, what tag/branch name would be meaningful for you? If we look 6 months from now, what name would make sense?

I'm after something that evokes legacy / impending deprecation :-)
Any suggestions? :-)

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+  - This objective implies remote signing.
+- Be golang native.
+  - Write documentation and examples for `godoc`.
+  - Test using `go test` with both positive and negative test cases.
+
+## Implementation Gaps
+
+Here are the gaps between the current implementation and the objectives:
+
+- The current implementation implements `COSE_Sign` and `COSE_Sign1`. However, the verification process does not follow [RFC8152](https://datatracker.ietf.org/doc/html/rfc8152) and causes [issue #7](https://github.com/veraison/go-cose/issues/7). There are also implementation errors in signing and verification (see [issue #8](https://github.com/veraison/go-cose/issues/8)).
```
Comment (yogeshbdeshpande) [inline DESIGN.md:20]: I think the tense of the sentence should now reflect past tense with the implementation, example like earlier implementation. It should also say, had caused issues like #7 and #8 which is been fixed in the current implementation.

File: DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+- The efficiency of the current implementation can be improved with crypto domain knowledge. For instances,
+  - Remove the use of constant time copy in `I2OSP`.
+  - Remove the check on `(r,s)` against `n`.
+- The current implementation implements `PS256`, `ES256`, `ES384`, `ES512`, and misses other common algorithms like `PS384`, `PS512`, `EdDSA` (e.g. `Ed25519`).
+  - All conventional signature schemes are _signature with appendix_.
+  - `RS256`, `RS384`, `RS512`, `ES256K` are marked **NOT RECOMMENDED** by [RFC8812](https://datatracker.ietf.org/doc/html/rfc8812).
+- The current implementation has solid `Signer` and `Verify` structures and can only sign or verify against certain algorithms (see [issue #9](https://github.com/veraison/go-cose/issues/9)). The implementation cannot be extended to have new algorithms or new implementation for existing algorithms.
+- The current implementation is not golang native.
```
Comment (yogeshbdeshpande) [inline DESIGN.md:28]: Is Line 28 referring to veraison/go-cose or some other repo? when it refers to golang nativity ? 

As far as I reckon that was with older notary implementation ? If so needs to be specified clearly.

**Comment by SteveLasker** [comment]
Agreed on the fork info atop. We can update the readme a bit later when we’ve decoupled. For now, just getting the code merged is a great start so we can start the dependencies and flush out the next round of issues

**Comment by qmuntal** [comment]
> > anch name would be meaningful for
> 
> I suggest a simple strategy, example(Feel free to shoot it down): go-cose-v.1.0.0
> 
> and then when Notary major release comes go-cose-v.2.0.0 2.0.1 for notary patches and 2.1.1 for notary enhancements?

Module tags should strictly follow [semantic versioning](https://semver.org/) with `v` prefix, else Go toolchain won't like our releases. This is what I would do:

- Tag main as-is now with `v0.1.0`.
- Merge this PR and create a new tag named `v0.2.0`.
- Once we pass all security reviews and we are confident we have a stable API, we can aim for the `v1.0.0`.

**Comment by yogeshbdeshpande** [comment]
> > > anch name would be meaningful for
> > 
> > 
> > I suggest a simple strategy, example(Feel free to shoot it down): go-cose-v.1.0.0
> > and then when Notary major release comes go-cose-v.2.0.0 2.0.1 for notary patches and 2.1.1 for notary enhancements?
> 
> Module tags should strictly follow [semantic versioning](https://semver.org/) with `v` prefix, else Go toolchain won't like our releases. This is what I would do:
> 
> * Tag main as-is now with `v0.1.0`.
> * Merge this PR and create a new tag named `v0.2.0`.
> * Once we pass all security reviews and we are confident we have a stable API, we can aim for the `v1.0.0`.

Works for me!

**Comment by shizhMSFT** [comment]
Golang projects follows [SemVer2](https://semver.org/).

> Given a version number MAJOR.MINOR.PATCH, increment the:
>
> 1. MAJOR version when you make incompatible API changes,
> 2. MINOR version when you add functionality in a backwards compatible manner, and
> 3. PATCH version when you make backwards compatible bug fixes.
>
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

The previous `go-cose` implementation is not SemVer2 versioned. To golang, it is still at `v0.0.0`. Then we can start with `v1.0.0`.

If the previous implementation is required to be `v1.0.0`, then we can have this implementation as `v2.0.0`. However, if we do want `v2.0.0`, we need to update the `go.mod` file to `github.com/veraison/go-cose/v2`. Meanwhile, we have to fix all vulnerabilities in `v1.0.0` or deprecate `v1.0.0` immediately after tagging it for security concerns.

**Comment by SteveLasker** [comment]
what about v0.0.0-sign1

**Comment by qmuntal** [comment]
> what about v0.0.0-sign1 , with comments that this is NOT a supported release with 1.0 as the supported codebase

Like it and +1 one what @shizhMSFT said. I would just not start with the 1.0 but with 0.1.0 so we can still iterate on the API a few weeks more without breaking any compatibility promise.

File: README.md (hunk 1 above)
Comment (SteveLasker) [reply README.md:3]: Suggest we manage the readme as a separate PR.

File: DESIGN.md (hunk 2 above)
Comment (SteveLasker) [reply DESIGN.md:20]: @shizhMSFT, can you make this minor wording change? 
Suggested change:

Prior versions of go-cose implemented `COSE_Sign` and `COSE_Sign1`. However, the verification process did not follow [RFC8152](https://datatracker.ietf.org/doc/html/rfc8152) which had caused issues like [#7](https://github.com/veraison/go-cose/issues/7) and [#8](https://github.com/veraison/go-cose/issues/8) which have been fixed in the current implementation.

File: DESIGN.md (hunk 3 above)
Comment (SteveLasker) [reply DESIGN.md:28]: @shizhMSFT, what do you think?

**Comment by SteveLasker** [comment]
I've created tag [v0.0.0-sign1-alpha.0](https://github.com/veraison/go-cose/releases/tag/v0.0.0-sign1-alpha.0) and created the branch [v0.0.0-sign1-alpha.0](https://github.com/veraison/go-cose/tree/v0.0.0-sign1-alpha.0)

I've also updated issue #37 to reflect the steps

@shizhMSFT, if you could make the minor changes noted above
- [line 20](https://github.com/veraison/go-cose/pull/51#discussion_r854137070)
- [line 28](https://github.com/veraison/go-cose/pull/51#discussion_r854143127)

I believe we're ready to merge

File: DESIGN.md (hunk 3 above)
Comment (shizhMSFT) [reply DESIGN.md]: updated in 6ef3dd4

File: DESIGN.md (hunk 2 above)
Comment (shizhMSFT) [reply DESIGN.md]: updated in 6ef3dd4

**Review by thomas-fossati — APPROVED** [review]
🚢 it!

**Review by yogeshbdeshpande — APPROVED** [review]

**Comment by yogeshbdeshpande** [comment]
Great Job Shiwei! Thank you for all the Hard Work.

**Review by SteveLasker — APPROVED** [review]
LGTM

**Pull request merged**